driver/src/main/java/org/neo4j/driver/internal/connector/socket/SSLSocketChannel.java

```diff
@@ -0,0 +1,493 @@
+     * in a call could be requiring more bytes to decipher, too small buffers to hold all data, etc.
+     *
+     * To verify if deciphering is done successfully, we could check if any bytes has been read into {@code buffer},
+     * as the deciphered bytes will only be saved into {@code buffer} when deciphering is carried out successfully.
+     *
+     * @param buffer
+     * @return
+     * @throws IOException
+     */
+    private HandshakeStatus unwrap( ByteBuffer buffer ) throws IOException
+    {
+        HandshakeStatus handshakeStatus = sslEngine.getHandshakeStatus();
+
+        /**
+         * This is the only place to read from the underlying channel
+         */
+        if ( channel.read( cipherIn ) < 0 )
+        {
+            throw new ClientException( "Failed to establish SSL socket connection." );
+        }
+        cipherIn.flip();
+
+        Status status = null;
+        do
+        {
+            status = sslEngine.unwrap( cipherIn, plainIn ).getStatus();
+            // Possible status here:
+            // OK - good
+            // BUFFER_OVERFLOW - we need to enlarge* plainIn
+            // BUFFER_UNDERFLOW - we need to enlarge* cipherIn and read more bytes from channel
+
+            /* In normal situations, enlarging buffers should happen very very rarely, as usually,
+             the initial size for application buffer and network buffer is greater than 1024 * 10,
+             and by using ChunkedInput and ChunkedOutput, the chunks are limited to 8192.
+             So we should not get any enlarging buffer request in most cases. */
+            switch ( status )
+            {
+            case OK:
+                plainIn.flip();
+                bufferCopy( plainIn, buffer );
+                plainIn.compact();
+                handshakeStatus = runDelegatedTasks();
+                break;
+            case BUFFER_OVERFLOW:
+                // Could attempt to drain the plainIn buffer of any already obtained
+                // data, but we'll just increase it to the size needed.
+                int curAppSize = plainIn.capacity();
+                int appSize = sslEngine.getSession().getApplicationBufferSize();
+                int newAppSize = appSize + plainIn.remaining();
+                if ( newAppSize > appSize * 2 )
+                {
+                    throw new ClientException(
+                            String.format( "Failed ro enlarge application buffer from %s to %s, as the maximum " +
+                                           "buffer size allowed is %s. The content in the buffer is: %s\n",
+                                    curAppSize, newAppSize, appSize * 2, BytePrinter.hex( plainIn ) ) );
+                }
+                ByteBuffer newPlainIn = ByteBuffer.allocateDirect( newAppSize );
+                newPlainIn.put( plainIn );
+                plainIn = newPlainIn;
+                logger.debug( "Enlarged application input buffer from %s to %s. " +
```

Comment:
I did not move this under `if( logger.debugEnabled() )` as we expect we will only get `OK` when invoking `wrap` and `unwrap`. If we got this debug info, it is probably an indication of a potential error in the code, e.g. forgot to clean the buffer after use etc.